File: packages/octocode-minifier/src/core/minifier.ts
import { getExtension } from '../utils/fileExtension.js';
import type { FileTypeMinifyConfig, MinifyResult } from '../types/index.js';
import { MINIFY_CONFIG, INDENTATION_SENSITIVE_NAMES } from '../types/index.js';
import {
  minifyConservativeCore,
  minifyAggressiveCore,
  minifyJsonCore,
  minifyGeneralCore,
  minifyMarkdownCore,
  minifyCSSCore,
  minifyHTMLCore,
  minifyJavaScriptCore,
  minifyTypeScriptLikeSync,
  minifyTypeScriptLike,
  minifyWithTerserSync,
  minifyWithTerser,
  minifyCSSAsync,
  minifyHTMLAsync,
  minifyComponentAsync,
} from './strategies.js';

const MINIFIER_EXT_OPTIONS = { lowercase: true, fallback: 'txt' } as const;
const TYPESCRIPT_LIKE_EXTENSIONS = new Set(['ts', 'tsx', 'jsx']);
const COMPONENT_EXTENSIONS = new Set(['vue', 'svelte']);

function getFileConfig(filePath: string): FileTypeMinifyConfig {
  const ext = getExtension(filePath, MINIFIER_EXT_OPTIONS);
  const baseName = (filePath.split(/[\\/]/).pop() || '').toLowerCase();

  if (INDENTATION_SENSITIVE_NAMES.has(baseName)) {
    return { strategy: 'conservative', comments: 'hash' };
  }

  return MINIFY_CONFIG.fileTypes[ext] || { strategy: 'general' };
}

function hasFlowSyntaxHint(content: string): boolean {
  return (
    content.includes('@flow') ||
    content.includes('import type ') ||
    content.includes('export type ') ||
    content.includes(' as any as ')
  );
}

function useTerserFallbackIfSupported(
  ext: string,
  content: string,
  fallbackContent: string
): string | null {
  if (ext !== 'jsx' && !(ext === 'js' && hasFlowSyntaxHint(content))) {
    return null;
  }

  return fallbackContent !== content && fallbackContent.length < content.length
    ? fallbackContent
    : null;
}

function shorterThanOriginal(
  content: string,
  candidate: string
): string | null {
  return candidate.length < content.length ? candidate : null;
}

function minifyTypeScriptLikeWithFallbackSync(
  content: string,
  filePath: string,
  fallbackContent: string
): string {
  const result = minifyTypeScriptLikeSync(content, filePath);
  const candidate = !result.failed
    ? shorterThanOriginal(content, result.content)
    : null;

  return candidate ?? fallbackContent;
}

async function minifyTypeScriptLikeWithFallback(
  content: string,
  filePath: string,
  fallbackContent: string
): Promise<{ content: string; reason?: string }> {
  const result = await minifyTypeScriptLike(content, filePath);
  const candidate = !result.failed
    ? shorterThanOriginal(content, result.content)
    : null;

  return {
    content: candidate ?? fallbackContent,
    ...(result.reason && { reason: result.reason }),
  };
}

export function minifyContentSync(content: string, filePath: string): string {
  const config = getFileConfig(filePath);
  const ext = getExtension(filePath, MINIFIER_EXT_OPTIONS);

  try {
    switch (config.strategy) {
      case 'terser':
        if (TYPESCRIPT_LIKE_EXTENSIONS.has(ext)) {
          return minifyTypeScriptLikeWithFallbackSync(
            content,
            filePath,
            minifyJavaScriptCore(content)
          );
        }
        return minifyWithTerserSync(content).content;

      case 'json':
        return minifyJsonCore(content).content;

      case 'conservative':
        if (TYPESCRIPT_LIKE_EXTENSIONS.has(ext)) {
          return minifyTypeScriptLikeWithFallbackSync(
            content,
            filePath,
            minifyConservativeCore(content, config)
          );
        }
        return minifyConservativeCore(content, config);

      case 'markdown':
        return minifyMarkdownCore(content);

      case 'aggressive':
        if (['css', 'less', 'scss'].includes(ext)) {
          return minifyCSSCore(content);
        }
        if (['html', 'htm', 'xml', 'svg'].includes(ext)) {
          return minifyHTMLCore(content);
        }
        return minifyAggressiveCore(content, config);

      case 'general':
      default:
        return minifyGeneralCore(content);
    }
  } /* v8 ignore start */ catch {
    return content;
  } /* v8 ignore stop */
}

export async function minifyContent(
  content: string,
  filePath: string
): Promise<MinifyResult> {
  try {
    const MAX_SIZE = 1024 * 1024;
    const contentSize = Buffer.byteLength(content, 'utf8');

    if (contentSize > MAX_SIZE) {
      return {
        content,
        failed: true,
        type: 'failed',
        reason: `File too large: ${(contentSize / 1024 / 1024).toFixed(2)}MB exceeds 1MB limit`,
      };
    }

    const config = getFileConfig(filePath);
    const ext = getExtension(filePath, MINIFIER_EXT_OPTIONS);

    switch (config.strategy) {
      case 'terser': {
        if (TYPESCRIPT_LIKE_EXTENSIONS.has(ext)) {
          const fallback = minifyJavaScriptCore(content);
          const result = await minifyTypeScriptLikeWithFallback(
            content,
            filePath,
            fallback
          );

          return {
            content: result.content,
            failed: false,
            type: 'terser',
            ...(result.reason && { reason: result.reason }),
          };
        }

        const result = await minifyWithTerser(content);
        if (result.failed) {
          const fallbackContent = useTerserFallbackIfSupported(
            ext,
            content,
            minifyJavaScriptCore(content)
          );
          if (fallbackContent !== null) {
            return {
              content: fallbackContent,
              failed: false,
              type: 'terser',
              ...(result.reason && { reason: result.reason }),
            };
          }
        }

        return {
          content: result.content,
          failed: result.failed,
          type: result.failed ? 'failed' : 'terser',
          ...(result.reason && { reason: result.reason }),
        };
      }

      case 'json': {
        const result = minifyJsonCore(content);
        // minifyJsonCore always returns failed:false and never sets reason;
        // the fields are kept explicit for future-proofing.
        return {
          content: result.content,
          failed: false,
          type: 'json',
        };
      }

      case 'conservative':
        if (TYPESCRIPT_LIKE_EXTENSIONS.has(ext)) {
          const fallback = minifyConservativeCore(content, config);
          const result = await minifyTypeScriptLikeWithFallback(
            content,
            filePath,
            fallback
          );

          return {
            content: result.content,
            failed: false,
            type: 'conservative',
            ...(result.reason && { reason: result.reason }),
          };
        }

        return {
          content: minifyConservativeCore(content, config),
          failed: false,
          type: 'conservative',
        };

      case 'general':
        return {
          content: minifyGeneralCore(content),
          failed: false,
          type: 'general',
        };

      case 'markdown':
        return {
          content: minifyMarkdownCore(content),
          failed: false,
          type: 'markdown',
        };

      case 'aggressive': {
        if (['css', 'less', 'scss'].includes(ext)) {
          const result = await minifyCSSAsync(content);
          return {
            content: result.content,
            failed: false,
            type: 'aggressive',
            ...(result.reason && { reason: result.reason }),
          };
        }

        if (['html', 'htm'].includes(ext)) {
          const result = await minifyHTMLAsync(content);
          return {
            content: result.content,
            failed: false,
            type: 'aggressive',
            ...(result.reason && { reason: result.reason }),
          };
        }

        if (COMPONENT_EXTENSIONS.has(ext)) {
          const result = await minifyComponentAsync(content);
          return {
            content: result.content,
            failed: false,
            type: 'aggressive',
            ...(result.reason && { reason: result.reason }),
          };
        }

        return {
          content: minifyAggressiveCore(content, config),
          failed: false,
          type: 'aggressive',
        };
      }
    }
  } catch (error: unknown) {
    return {
      content,
      failed: true,
      type: 'failed',
      reason: `Unexpected minification error: ${error instanceof Error ? error.message : 'Unknown error'}`,
    };
  }
}

export { MINIFY_CONFIG };

File: packages/octocode-minifier-utils/src/minifier.rs
use crate::config::{minify_config, indentation_sensitive_names, FileTypeConfig};
use crate::file_extension::get_extension_internal;
use crate::types::MinifyResult;
use crate::strategies::{
    minify_conservative, minify_aggressive,
    minify_general_core, minify_markdown_core,
    minify_css_quality,
    minify_html_core, minify_html_quality,
    minify_javascript_core, minify_js_oxc,
    minify_json_core_inner,
};

pub(crate) const MAX_SIZE: usize = 1024 * 1024; // 1 MB guard, shared by all FFI content entry points

pub fn get_file_config(file_path: &str) -> Option<&'static FileTypeConfig> {
    let ext = get_extension_internal(file_path, true, "txt");
    let basename = file_path
        .rsplit(['/', '\\'])
        .next()
        .unwrap_or(file_path)
        .to_lowercase();

    if indentation_sensitive_names().contains(basename.as_str()) {
        return minify_config().get("sh"); // hash comments, conservative
    }

    minify_config().get(ext.as_str())
}

pub fn comment_groups(cfg: &FileTypeConfig) -> Vec<&'static str> {
    cfg.comments.map(|c| c.to_vec()).unwrap_or_default()
}

/// Synchronous full minification — mirrors TS `minifyContentSync`.
pub fn minify_content_sync_inner(content: &str, file_path: &str) -> String {
    if content.len() > MAX_SIZE { return content.to_owned(); }
    dispatch(content, file_path, false)
}

/// Full minification returning MinifyResult — mirrors TS `minifyContent` (async in TS).
/// In Rust all operations are sync; callers may wrap in Promise.resolve() on the JS side.
pub fn minify_content_result_inner(content: &str, file_path: &str) -> MinifyResult {
    let content_size = content.len();
    if content_size > MAX_SIZE {
        return MinifyResult::fail(
            content.to_owned(),
            format!("File too large: {:.2}MB exceeds 1MB limit", content_size as f64 / 1_048_576.0),
        );
    }

    let Some(cfg) = get_file_config(file_path) else {
        let out = minify_general_core(content);
        return MinifyResult::ok(out, "general");
    };
    let ext  = get_extension_internal(file_path, true, "txt");
    let grps = comment_groups(cfg);

    let (out, strategy) = match cfg.strategy {
        "terser" | "conservative" => {
            let s = if matches!(ext.as_str(), "ts"|"tsx"|"js"|"jsx"|"mjs"|"cjs") {
                // OXC: full compression with mangle for the minify path
                minify_js_oxc(content, file_path, true)
                    .unwrap_or_else(|| minify_javascript_core(content))
            } else {
                minify_conservative(content, Some(&grps))
            };
            (s, cfg.strategy)
        }
        "aggressive" => {
            let s = if matches!(ext.as_str(), "css"|"less"|"scss") {
                minify_css_quality(content)          // lightningcss
            } else if matches!(ext.as_str(), "html"|"htm") {
                minify_html_quality(content)         // minify-html
            } else if matches!(ext.as_str(), "xml"|"svg") {
                minify_html_core(content)
            } else {
                minify_aggressive(content, Some(&grps))
            };
            (s, "aggressive")
        }
        "json" => {
            let (s, _) = minify_json_core_inner(content);
            (s, "json")
        }
        "markdown" => (minify_markdown_core(content), "markdown"),
        _           => (minify_general_core(content), "general"),
    };

    MinifyResult::ok(out, strategy)
}

// Shared dispatch (sync path without MinifyResult overhead)
fn dispatch(content: &str, file_path: &str, _high_quality: bool) -> String {
    let Some(cfg) = get_file_config(file_path) else {
        return minify_general_core(content);
    };
    let ext  = get_extension_internal(file_path, true, "txt");
    let grps = comment_groups(cfg);

    match cfg.strategy {
        "terser" | "conservative" => {
            if matches!(ext.as_str(), "ts"|"tsx"|"js"|"jsx"|"mjs"|"cjs") {
                minify_js_oxc(content, file_path, true)
                    .unwrap_or_else(|| minify_javascript_core(content))
            } else {
                minify_conservative(content, Some(&grps))
            }
        }
        "aggressive" => {
            if matches!(ext.as_str(), "css"|"less"|"scss") {
                minify_css_quality(content)
            } else if matches!(ext.as_str(), "html"|"htm") {
                minify_html_quality(content)
            } else if matches!(ext.as_str(), "xml"|"svg") {
                minify_html_core(content)
            } else {
                minify_aggressive(content, Some(&grps))
            }
        }
        "json"     => { let (s, _) = minify_json_core_inner(content); s }
        "markdown" => minify_markdown_core(content),
        _          => minify_general_core(content),
    }
}
